# Задание 1. Загрузка данных
Изучить представленный набор данных на основе описания его столбцов, загрузить его и оставить 8 столбцов для дальнейшего изучения: surgery?, Age, rectal temperature, pulse, respiratory rate, temperature of extremities, pain, outcome.

In [2]:
# читаем файл 
import pandas as pd
import numpy as np
horse_data = pd.read_csv('horse_data.csv', sep=',', header = None, na_values = '?') 
# header = None, иначе в кач-ве заголовка читается 1 строка с данными
# na_values = '?' - сразу заменяем ?, обозн. пропущ значения, на NaN
horse_data

0   1        2     3      4     5    6    7    8    9   ...    18    19  \
0    2.0   1   530101  38.5   66.0  28.0  3.0  3.0  NaN  2.0  ...  45.0   8.4   
1    1.0   1   534817  39.2   88.0  20.0  NaN  NaN  4.0  1.0  ...  50.0  85.0   
2    2.0   1   530334  38.3   40.0  24.0  1.0  1.0  3.0  1.0  ...  33.0   6.7   
3    1.0   9  5290409  39.1  164.0  84.0  4.0  1.0  6.0  2.0  ...  48.0   7.2   
4    2.0   1   530255  37.3  104.0  35.0  NaN  NaN  6.0  2.0  ...  74.0   7.4   
..   ...  ..      ...   ...    ...   ...  ...  ...  ...  ...  ...   ...   ...   
295  1.0   1   533886   NaN  120.0  70.0  4.0  NaN  4.0  2.0  ...  55.0  65.0   
296  2.0   1   527702  37.2   72.0  24.0  3.0  2.0  4.0  2.0  ...  44.0   NaN   
297  1.0   1   529386  37.5   72.0  30.0  4.0  3.0  4.0  1.0  ...  60.0   6.8   
298  1.0   1   530612  36.5  100.0  24.0  3.0  3.0  3.0  1.0  ...  50.0   6.0   
299  1.0   1   534618  37.2   40.0  20.0  NaN  NaN  NaN  NaN  ...  36.0  62.0   

      20   21   22  23     24  25  26  27  
0    NaN  NaN  2.0   2  11300   0   0   2  
1    2.0  2.0  3.0   2   2208   0   0   2  
2    NaN  NaN  1.0   2      0   0   0   1  
3    3.0  5.3  2.0   1   2208   0   0   1  
4    NaN  NaN  2.0   2   4300   0   0   2  
..   ...  ...  ...  ..    ...  ..  ..  ..  
295  NaN  NaN  3.0   2   3205   0   0   2  
296  3.0  3.3  3.0   1   2208   0   0   1  
297  NaN  NaN  2.0   1   3205   0   0   2  
298  3.0  3.4  1.0   1   2208   0   0   1  
299  1.0  1.0  3.0   2   6112   0   0   2  

[300 rows x 28 columns]

In [3]:
# формируем датафрейм с нужным набором столбцов - индексы смещены на -1 относительно нумерации в текст.файле
data = horse_data.loc[:, [0, 1, 3, 4, 5, 6, 10, 22]]
# помимо нужных параметров, оставила еще hospital number - id лошадей, теоретически, он может пригодиться в дальнейшей работе
# переименовываем столбцы
data.columns = ['surgery', 'age', 'rect_temp', 'pulse', 'resp_rate', 'extr_temp', 'pain', 'outcome']
data.head()

surgery  age  rect_temp  pulse  resp_rate  extr_temp  pain  outcome
0      2.0    1       38.5   66.0       28.0        3.0   5.0      2.0
1      1.0    1       39.2   88.0       20.0        NaN   3.0      3.0
2      2.0    1       38.3   40.0       24.0        1.0   3.0      1.0
3      1.0    9       39.1  164.0       84.0        4.0   2.0      2.0
4      2.0    1       37.3  104.0       35.0        NaN   NaN      2.0

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
surgery      299 non-null float64
age          300 non-null int64
rect_temp    240 non-null float64
pulse        276 non-null float64
resp_rate    242 non-null float64
extr_temp    244 non-null float64
pain         245 non-null float64
outcome      299 non-null float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


# Задание 2. Первичное изучение данных    
Проанализировать значения по столбцам, рассчитать базовые статистики, найти выбросы.

В нашем датафрейме 2 типа данных:  
**Качественные** - 'surgery', 'age', 'extr_temp', 'pain', 'outcome' - для них можно найти только моду. Заодно можно посмотреть долю модального значения в общем объеме данных, чтобы понимать, насколько оно распространено.  
**Количественные** - 'rect_temp', 'pulse', 'resp_rate' - для них можно рассчитать все базовые статистики.

### Сначала посмотрим качественные данные  

In [5]:
def quality_data_stat(param, quantity):
    print('Модальное значение: ', int(param.value_counts().idxmax()))   # то же самое, что и .mode(), только без вывода типа переменной
    print('Доля, %: ', round((param.value_counts().max() / quantity * 100), 1))

quantity = 300
print('surgery - была ли операция')
quality_data_stat(data['surgery'], quantity)

surgery - была ли операция
Модальное значение:  1
Доля, %:  60.0


Бедных больных лошадок чаще резали, чем не резали. Т.е. в 60 % случаев было необходимо оперативное вмешательство. 

In [6]:
print('age - возраст лошади')
quality_data_stat(data['age'], quantity)

age - возраст лошади
Модальное значение:  1
Доля, %:  92.0


Практически вся выборка в рамках исследования - это взрослые лошади.   
Надо отметить, что в расшифровке показателей второе значение категории - "2", а в данных - "9". Скорее всего это опечатка, потому что иначе непонятно, зачем было вводить разделение на взрослых лошадей и жеребят. 

In [7]:
print('extr_temp - экспертная оценка температуры конечностей')
quality_data_stat(data['extr_temp'], quantity)

extr_temp - экспертная оценка температуры конечностей
Модальное значение:  3
Доля, %:  36.3


Наиболее распространенная оценка - "3". Она соответствует пониженной температуре лошадиных конечностей и говорит о том, что лошадки чаще всего чувствуют себя не очень здорово, но и не не совсем плохо. Ведь в данных есть еще оценка "4" - ледяные лапочки как индикатор возможного болевого шока. 

In [8]:
print('pain - экспертная оценка болевых ощущений')
quality_data_stat(data['pain'], quantity)

pain - экспертная оценка болевых ощущений
Модальное значение:  3
Доля, %:  22.3


Чаще всего встречается оценка "3" - периодическая слабая боль. Но вообще доля этого значения не такая уж большая, так что в выборке наверняка есть другие близкие по частоте появления варианты. Использовать эту моду для характеристики всех данных я бы не стала. 

In [9]:
print('outcome - результаты')
quality_data_stat(data['outcome'], quantity)

outcome - результаты
Модальное значение:  1
Доля, %:  59.3


Радует, что лошадки все-таки чаще выживали, чем умирали :)

## Базовые статистические показатели для количественных данных

In [10]:
# Для удобства упаковываем все необходимые расчеты в функцию
def statistics(param):

    print('Среднее значение: ', param.round().mean())
    print('Мода: ', param.round().mode()[0])
    print('Медиана: ', param.median())
    print('Максимальное значение: ', param.max())
    print('Минимальное значение: ', param.min())
    print('Размах: ', (param.max() - param.min()))
    print('Среднее квадратическое отклонение: ', param.std())
    print('Дисперсия: ', param.var())
    Q1 = param.quantile(0.25)
    Q3 = param.quantile(0.75)
    print('Q1 - первый квартиль: ', Q1)
    print('Q3 - третий квартиль: ', Q3)
    print('Межквартильный размах: ', (Q3-Q1))

# Подставляем в функцию данные и печатаем с необходимыми пояснениями
print('Ректальная температура', '\n', 'Норма для взрослой лошади - 37.8 С', '\n','-----')
statistics(data['rect_temp'])

Ректальная температура 
 Норма для взрослой лошади - 37.8 С 
 -----
Среднее значение:  38.1375
Мода:  38.0
Медиана:  38.2
Максимальное значение:  40.8
Минимальное значение:  35.4
Размах:  5.399999999999999
Среднее квадратическое отклонение:  0.7322886641121578
Дисперсия:  0.5362466875871686
Q1 - первый квартиль:  37.8
Q3 - третий квартиль:  38.5
Межквартильный размах:  0.7000000000000028


Какой бы метод подсчета среднего значения мы ни использовали, мы получаем среднюю температуру по больнице. И эта температура соответствует нормам для здоровой взрослой лошади - от 37,8 до 38,4 (если верить некоторым интернет-источникам).  
Так что средние в данном конкретном случае особого смысла не имеют, потому что живые организмы могут оставаться живыми при очень ограниченном диапазоне температуры тела, и критичное отклонение от нормы на пару градусов в "+" или "-" при болезнях случается, видимо, одинаково часто. 

In [11]:
print('Пульс', '\n', 'Норма для взрослой лошади - 30 - 40 ударов/мин', '\n','-----')
statistics(data['pulse'])

Пульс 
 Норма для взрослой лошади - 30 - 40 ударов/мин 
 -----
Среднее значение:  71.91304347826087
Мода:  48.0
Медиана:  64.0
Максимальное значение:  184.0
Минимальное значение:  30.0
Размах:  154.0
Среднее квадратическое отклонение:  28.630556660735003
Дисперсия:  819.7087747035575
Q1 - первый квартиль:  48.0
Q3 - третий квартиль:  88.0
Межквартильный размах:  40.0


Средние значения пульса различаются очень сильно, что говорит о большой неоднородности данных. Возможно, в этой ситуации имеет смысл ориентироваться не на их абсолютные значения, а на то, как все данные соотносятся между собой.  
Минимальное значение соответствует нормальному пульсу взрослой лошади в спокойном состоянии.  
Чаще других встречалось значение, чуть превышающее норму (но надо смотреть на долю таких значений, их может быть, например, всего 3 или 5, просто у нас очень большой интервал возможных значений и другие встречались 1-2 раза). 
Медиана превышает норму для здорового животного в 1,5 - 2 раза. И все равно выше нее лежат еще 50 % наблюдений. 
Максимальное значение - 184, что, в среднем, в 5 раз выше нормы. 
Соответственно, у нас очень большие показатели размаха и дисперсии.   
В целом, у больных лошадей сердцебиение может ускоряться до запредельных величин, но не может опускаться ниже нормальных значений здорового животного. 


In [12]:
print('Количество вдохов в минуту', '\n', 'Норма для взрослой лошади - 8 - 10 вдохов/мин', '\n','-----')
statistics(data['resp_rate'])

Количество вдохов в минуту 
 Норма для взрослой лошади - 8 - 10 вдохов/мин 
 -----
Среднее значение:  30.417355371900825
Мода:  20.0
Медиана:  24.5
Максимальное значение:  96.0
Минимальное значение:  8.0
Размах:  88.0
Среднее квадратическое отклонение:  17.642231385134664
Дисперсия:  311.24832824663054
Q1 - первый квартиль:  18.5
Q3 - третий квартиль:  36.0
Межквартильный размах:  17.5


С частотой дыхания точно такая же ситуация, как и с пульсом. Минимальное значение - норма для здорового животного, максимальное значение превышает норму в 10 раз. Видимо, чем хуже чувствует себя больное животное, тем больше вдохов в минуту оно делает. Это дает большой разброс средних значений и большие показатели размаха и дисперсии. И вопрос не в конкретных числовых значениях средних величин, а в том, каким вариантам развития событий они соответствуют (т.е. надо смотреть на взаимосвязь с другими переменными). 

## Выбросы

**Сразу напишу выводы: я за то, чтобы оставить все выбросы.   
Есть подозрение, что основной смысл сбора и анализа этих данных - это не найти какие-то усредненные нормальные значения, а исследовать все многообразие отклонений от нормы, чтобы понять, насколько тяжело могут болеть бедные лошадки и к каким исходам это может привести.** 

### Работа с выбросами для ректальной температуры

In [13]:
# Находим верхнюю и нижнюю границу  
q1 = data['rect_temp'].quantile(0.25)
q3 = data['rect_temp'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
print('Нижняя граница: ', lower_bound)
print('Верхняя граница: ', upper_bound)

# Находим датафрейм без выбросов
# Заодно выпадают и строки с пустыми значениями (они не влияют на расчет средних). Помним о них но пока игнорируем
remove_outliers_rt = data[data['rect_temp'].between(lower_bound, upper_bound, inclusive=True)].sort_values('rect_temp')
remove_outliers_rt

Нижняя граница:  36.74999999999999
Верхняя граница:  39.550000000000004


surgery  age  rect_temp  pulse  resp_rate  extr_temp  pain  outcome
113      1.0    1       36.8   60.0       28.0        NaN   NaN      2.0
277      2.0    1       36.9   50.0       40.0        2.0   1.0      1.0
70       2.0    1       37.0   60.0       20.0        3.0   3.0      3.0
292      1.0    1       37.0   66.0       20.0        1.0   4.0      2.0
190      1.0    1       37.1   40.0        8.0        NaN   3.0      1.0
..       ...  ...        ...    ...        ...        ...   ...      ...
165      1.0    1       39.4  120.0       48.0        NaN   NaN      3.0
231      2.0    1       39.5   92.0       28.0        3.0   5.0      2.0
287      2.0    9       39.5   84.0       30.0        NaN   NaN      1.0
117      2.0    1       39.5    NaN        NaN        3.0   3.0      3.0
162      2.0    1       39.5   60.0       10.0        3.0   3.0      1.0

[226 rows x 8 columns]

In [14]:
# Теперь посмотрим, насколько отличаются расчет статистик для общей базы и данных без выбросов
print('Среднее без выбросов: ', remove_outliers_rt['rect_temp'].mean())
print('Медиана без выбросов: ', remove_outliers_rt['rect_temp'].median())
print('Разница между средним и средним без выбросов: ', data['rect_temp'].mean() - remove_outliers_rt['rect_temp'].mean())
print('Разница между медианой и медианой без выбросов: ', data['rect_temp'].median() - remove_outliers_rt['rect_temp'].median())

Среднее без выбросов:  38.169026548672555
Медиана без выбросов:  38.2
Разница между средним и средним без выбросов:  -0.0011098820058634828
Разница между медианой и медианой без выбросов:  0.0


Выбросы практически не влияют на средние значения. Поэтому их можно оставить. Тем более, если более подробно изучить эти самые выбросы (см. ниже), то окажется, что их всего 14, из них половина выпадает за нижнюю границу и половина - за верхнюю.

In [15]:
# Ищем строки с выбросами (сливаем датафреймы, удаляем дубликаты и строки с пустыми значениями по столбцам)
rect_temp_outl = pd.concat([data, remove_outliers_rt]).drop_duplicates(keep=False)
rect_temp_outl = rect_temp_outl[rect_temp_outl['rect_temp'].notna()]
rect_temp_outl

surgery  age  rect_temp  pulse  resp_rate  extr_temp  pain  outcome
20       1.0    1       39.9   72.0       60.0        1.0   5.0      1.0
44       1.0    1       35.4  140.0       24.0        3.0   4.0      3.0
54       2.0    1       40.3  114.0       36.0        3.0   2.0      3.0
75       1.0    9       39.7  100.0        NaN        3.0   2.0      3.0
80       1.0    1       36.4   98.0       35.0        3.0   4.0      2.0
91       2.0    1       40.3  114.0       36.0        3.0   2.0      2.0
99       2.0    1       39.6  108.0       51.0        3.0   2.0      1.0
118      1.0    1       36.5   78.0       30.0        1.0   5.0      1.0
141      2.0    1       36.0   42.0       30.0        NaN   NaN      2.0
238      2.0    1       36.1   88.0        NaN        3.0   3.0      3.0
251      2.0    1       36.6   42.0       18.0        3.0   1.0      2.0
259      1.0    1       40.8   72.0       42.0        3.0   2.0      2.0
281      2.0    1       40.0   78.0        NaN        3.0   2.0      2.0
298      1.0    1       36.5  100.0       24.0        3.0   3.0      1.0

### Выбросы для пульса 

In [16]:
# Находим верхнюю и нижнюю границу  
q1 = data['pulse'].quantile(0.25)
q3 = data['pulse'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
print('Нижняя граница: ', lower_bound)
print('Верхняя граница: ', upper_bound)

# Находим датафрейм без выбросов
# Заодно выпадают и строки с пустыми значениями (они не влияют на расчет средних). Помним о них но пока игнорируем
remove_outliers_pulse = data[data['pulse'].between(lower_bound, upper_bound, inclusive=True)].sort_values('pulse')
remove_outliers_pulse

Нижняя граница:  -12.0
Верхняя граница:  148.0


surgery  age  rect_temp  pulse  resp_rate  extr_temp  pain  outcome
130      1.0    1       37.8   30.0       12.0        NaN   NaN      2.0
232      1.0    1       38.5   30.0       18.0        NaN   NaN      1.0
203      2.0    1       37.2   36.0        9.0        1.0   2.0      1.0
276      1.0    1       38.0   36.0       16.0        3.0   4.0      3.0
242      1.0    1        NaN   36.0       32.0        1.0   5.0      2.0
..       ...  ...        ...    ...        ...        ...   ...      ...
148      1.0    1       38.3  132.0        NaN        NaN   2.0      1.0
135      1.0    9       38.1  136.0       48.0        3.0   5.0      2.0
44       1.0    1       35.4  140.0       24.0        3.0   4.0      3.0
103      1.0    9       38.0  140.0       68.0        1.0   3.0      1.0
39       1.0    9       39.2  146.0       96.0        NaN   NaN      2.0

[271 rows x 8 columns]

Нижняя граница ушла куда-то в глубокий минус. Видимо, из-за того, что у бедных больных лошадок сердце колотится так сильно, что иногда кол-во ударов в минуту в несколько раз превышает норму. Для нас это означает, что выбросов снизу не будет. 

In [17]:
# Теперь посмотрим, насколько отличаются раасчет статистик для общей базы и данных без выбросов
print('Среднее без выбросов: ', remove_outliers_pulse['pulse'].mean())
print('Медиана без выбросов: ', remove_outliers_pulse['pulse'].median())
print('Разница между средним и средним без выбросов: ', data['pulse'].mean() - remove_outliers_pulse['pulse'].mean())
print('Разница между медианой и медианой без выбросов: ', data['pulse'].median() - remove_outliers_pulse['pulse'].median())

Среднее без выбросов:  70.25830258302582
Медиана без выбросов:  64.0
Разница между средним и средним без выбросов:  1.6547408952350509
Разница между медианой и медианой без выбросов:  0.0


В данном случае я бы тоже оставила выбросы. Разницы между медианами нет, некоторая разница между средним и средним без выбросов есть, но обе эти величины нужно сравнивать не между собой, а с нормой для здоровых лошадей - а это 30 - 40 ударов/мин. Таким образом, и с выбросами, и без них, получается один и тот же вывод: средние значения пульса по выборке примерно в 2 раза превышают норму.  
А если мы посмотрим на эти самые выбросы в данных, то увидим, что это было, в среднем, 4-кратное превышение нормы и почти все эти лошадки в итоге погибли. И пусть хоть такая память о них останется.

In [18]:
# Ищем строки с выбросами (сливаем датафреймы, удаляем дубликаты и строки с пустыми значениями по столбцам)
pulse_outl = pd.concat([data, remove_outliers_pulse]).drop_duplicates(keep=False)
pulse_outl = pulse_outl[pulse_outl['pulse'].notna()]
pulse_outl

surgery  age  rect_temp  pulse  resp_rate  extr_temp  pain  outcome
3        1.0    9       39.1  164.0       84.0        4.0   2.0      2.0
41       2.0    9       39.0  150.0       72.0        NaN   NaN      1.0
55       1.0    9       38.6  160.0       20.0        3.0   3.0      2.0
255      1.0    9       38.8  184.0       84.0        1.0   4.0      2.0
275      1.0    9       38.8  150.0       50.0        1.0   5.0      2.0

### Выбросы для частоты дыхания 

In [19]:
# Находим верхнюю и нижнюю границу  
q1 = data['resp_rate'].quantile(0.25)
q3 = data['resp_rate'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
print('Нижняя граница: ', lower_bound)
print('Верхняя граница: ', upper_bound)

# Находим датафрейм без выбросов
# Заодно выпадают и строки с пустыми значениями (они не влияют на расчет средних). Помним о них но пока игнорируем
remove_outliers_r_rate = data[data['resp_rate'].between(lower_bound, upper_bound, inclusive=True)].sort_values('resp_rate')
remove_outliers_r_rate

Нижняя граница:  -7.75
Верхняя граница:  62.25


surgery  age  rect_temp  pulse  resp_rate  extr_temp  pain  outcome
190      1.0    1       37.1   40.0        8.0        NaN   3.0      1.0
203      2.0    1       37.2   36.0        9.0        1.0   2.0      1.0
140      1.0    1       38.1   44.0        9.0        3.0   2.0      1.0
89       2.0    1       38.3   42.0       10.0        1.0   1.0      1.0
162      2.0    1       39.5   60.0       10.0        3.0   3.0      1.0
..       ...  ...        ...    ...        ...        ...   ...      ...
220      1.0    1       39.2   88.0       58.0        4.0   5.0      3.0
20       1.0    1       39.9   72.0       60.0        1.0   5.0      1.0
23       1.0    9       38.3  130.0       60.0        NaN   2.0      1.0
185      1.0    1       38.5  120.0       60.0        4.0   NaN      1.0
248      1.0    1       38.0  108.0       60.0        2.0   4.0      1.0

[225 rows x 8 columns]

И опять у бедных больных лошадок нижняя граница ушла в минус, а верхняя - превышает норму более чем в 6 раз.

In [20]:
# Теперь посмотрим, насколько отличаются раасчет статистик для общей базы и данных без выбросов
print('Среднее без выбросов: ', remove_outliers_r_rate['resp_rate'].mean())
print('Медиана без выбросов: ', remove_outliers_r_rate['resp_rate'].median())
print('Разница между средним и средним без выбросов: ', data['resp_rate'].mean() - remove_outliers_r_rate['resp_rate'].mean())
print('Разница между медианой и медианой без выбросов: ', data['resp_rate'].median() - remove_outliers_r_rate['resp_rate'].median())

Среднее без выбросов:  26.715555555555557
Медиана без выбросов:  24.0
Разница между средним и средним без выбросов:  3.701799816345268
Разница между медианой и медианой без выбросов:  0.5


И опять я бы оставила выбросы. Если сравнивать средние с выбросами и без них не между собой, а с нормальным значением (8-10 вдохов/мин), то, по-любому, идет превышение нормы в 2,5 раза. Так что эти выбросы - еще одно подтверждение того, что лошадкам может быть очень и очень плохо. 
А еще, есть подозрение, что это достаточно существенный кусок данных - 17 историй болезни лошадок (5 % общей выборки или 7 % выборки без пропущенных значений по этому показателю (% считаем от 300 и от 242 соответственно)). 

In [21]:
# Ищем строки с выбросами (сливаем датафреймы, удаляем дубликаты и строки с пустыми значениями по столбцам)
resp_rate_outl = pd.concat([data, remove_outliers_r_rate]).drop_duplicates(keep=False)
resp_rate_outl = resp_rate_outl[resp_rate_outl['resp_rate'].notna()]
resp_rate_outl

surgery  age  rect_temp  pulse  resp_rate  extr_temp  pain  outcome
3        1.0    9       39.1  164.0       84.0        4.0   2.0      2.0
39       1.0    9       39.2  146.0       96.0        NaN   NaN      2.0
41       2.0    9       39.0  150.0       72.0        NaN   NaN      1.0
82       1.0    9       38.1  100.0       80.0        3.0   3.0      1.0
84       1.0    1       37.8   60.0       80.0        1.0   2.0      1.0
103      1.0    9       38.0  140.0       68.0        1.0   3.0      1.0
106      1.0    1       38.3   52.0       96.0        NaN   NaN      1.0
120      1.0    1       39.4   54.0       66.0        1.0   2.0      1.0
125      1.0    1       38.0   42.0       68.0        4.0   3.0      1.0
208      1.0    1       37.8   88.0       80.0        3.0   NaN      3.0
229      1.0    9       38.5  120.0       70.0        NaN   NaN      1.0
244      1.0    9       38.2  124.0       88.0        1.0   2.0      1.0
255      1.0    9       38.8  184.0       84.0        1.0   4.0      2.0
265      2.0    1        NaN   56.0       68.0        3.0   3.0      3.0
295      1.0    1        NaN  120.0       70.0        4.0   2.0      3.0

# Задание 3. Работа с пропусками    
Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе заполнения пропусков по каждому столбцу на основе рассчитанных статистик и возможной взаимосвязи значений в них. Сформировать датафрейм, в котором пропуски будут отсутствовать.

In [22]:
# смотрим на объем пропущенных данных 
(data.isna().mean() * 100).round(2)

surgery       0.33
age           0.00
rect_temp    20.00
pulse         8.00
resp_rate    19.33
extr_temp    18.67
pain         18.33
outcome       0.33
dtype: float64

У нас есть данные, по которым практически нет пропусков, и есть показатели с большим объемом пропущенной информации. Можно предположить, что все происходило примерно таким образом:   
Когда на больную лошадку заводили историю болезни, часть информации в эту самую историю можно было внести легко и просто - поставить галочку напротив нужного пункта: возраст, была операция или нет, лошадка умерла или выжила. Именно по таким фактическим показателям пропусков почти нет.   
А для получения остальных данных нужно было потратить время и приложить усилия: осмотреть животное, оценить его состояние, измерить пульс, температуру и т.д., внести все результаты в мед.карту в нужном формате.   
Скорее всего, не всегда и не все измерения и оценки проводились (или фиксировались) в полном объеме. Возможно, не всегда и не во всем была необходимость, ведь хороший специалист вполне способен оценить состояние животного при быстром осмотре. Или на это иногда не оставалось времени. Или в ту смену работал лаборант-пофигист. Или, например, кто-то из врачей в принципе игнорировал необходимость подробного заполнения истории болезни (и, возможно, к лечению лошадок подходил так же небрежно)...  и т.д. О причинах мы можем только гадать. Но, не зная точной причины пропусков, нельзя сделать более менее достоверные предположения о том, насколько данные с пропущенными результатами осмотра похожи(или отличаются) от полностью заполненных данных. Можем ли мы заполнять пропуски, на основе имеющихся значений и взаимосвязей между ними, или нам нельзя это делать из-за неучтенных факторов?   
**В общем, в данном конкретном случае я бы оставила NaN значения или попробовала бы поднять дополнительную информацию по исследованию, а так очень велик риск просто сфальсифицировать научные данные, учитывая небольшое число наблюдений и почти 20 % пропущенных значений по некоторым параметрам.**

Но у нас учебная задача - как-нибудь заполнить пропуски, поэтому, наверное, можно действовать так:  
  
**Начнем с данных с качественных данных**
По одному пропуску в "surgery" и "outcome" заменим модальными значениями. Будем считать, что при заполнении что-то напутали (причем в одной и той же строке). Если бы мы загружали номера мед.карт - можно было бы проверить, не дублируется ли эта строка с еще какой-нибудь в датафрейме. А так будем просто считать, что дублей нет. 
Пропуски в остальных категориальных данных заменим на 'no info', так как не знаем, почему они не были заполнены (см. причины выше или, например, состояние лошадки было нестабильным и сделать точную оценку было сначала затруднительно, потом бессмысленно). 

In [237]:
# строка с пропусками в операциях и исходах
data[pd.isnull(data['surgery'])]

surgery  age  rect_temp  pulse  resp_rate  extr_temp  pain  outcome
132      NaN    1       38.0   48.0       20.0        3.0   4.0      NaN

In [67]:
# скопируем рабочий датафрейм
data_without_nan = data.copy()
# заменим пропуски
data_without_nan['surgery'].fillna(data['surgery'].mode()[0], inplace=True)
data_without_nan['outcome'].fillna(data['outcome'].mode()[0], inplace=True)
data_without_nan['extr_temp'].fillna('no_info', inplace=True)
data_without_nan['pain'].fillna('no_info', inplace=True)
data_without_nan.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
surgery      300 non-null float64
age          300 non-null int64
rect_temp    240 non-null float64
pulse        276 non-null float64
resp_rate    242 non-null float64
extr_temp    300 non-null object
pain         300 non-null object
outcome      300 non-null float64
dtypes: float64(5), int64(1), object(2)
memory usage: 18.9+ KB


**Теперь перейдем к количественным данным**  
Будем заполнять пропуски с группировкой по достоверно известным нам переменным без пропусков - surgery и outcome. Возраст не берем, т.к. почти все лошади в исследовании были взрослыми, различий в категории нет, анализировать там нечего. 
И заполнять будем средним арифметическим (Теоретически, если в данных есть выбросы, лучше брать медиану. Практически в данном конкретном исследовании наличие большого числа выбросов я бы считала одной из характерных особенностей данных.) 

**Температура**  
Сначала группируем данные


In [70]:
print(data.groupby(['surgery', 'outcome'])['rect_temp'].mean())
print(data.groupby(['surgery'])['rect_temp'].mean())
print(data.groupby(['outcome'])['rect_temp'].mean())

surgery  outcome
1.0      1.0        38.146250
         2.0        38.187179
         3.0        38.052381
2.0      1.0        38.240278
         2.0        38.157143
         3.0        38.053846
Name: rect_temp, dtype: float64
surgery
1.0    38.143571
2.0    38.204040
Name: rect_temp, dtype: float64
outcome
1.0    38.190789
2.0    38.179245
3.0    38.052941
Name: rect_temp, dtype: float64


Никакой явной зависимости нет, все в пределах погрешности измерений градусников. Поэтому, наверное, нет особой разницы, как именно заполнять. Пусть это будет группировка по двум переменным.   
И по итогам заполнения получаем почти все то же самое, с чуть меньшим разбросом и вокруг среднего.

In [71]:
data_without_nan['rect_temp'].fillna(data.groupby(['surgery', 'outcome'])['rect_temp'].transform('mean'), inplace=True)

In [72]:
print('Ректальная температура до преобразований')
statistics(data['rect_temp'])
print('-------')
print('Температура после заполнения пропусков')
statistics(data_without_nan['rect_temp'])

Ректальная температура до преобразований
Среднее значение:  38.1375
Мода:  38.0
Медиана:  38.2
Максимальное значение:  40.8
Минимальное значение:  35.4
Размах:  5.399999999999999
Среднее квадратическое отклонение:  0.7322886641121578
Дисперсия:  0.5362466875871686
Q1 - первый квартиль:  37.8
Q3 - третий квартиль:  38.5
Межквартильный размах:  0.7000000000000028
-------
Температура после заполнения пропусков
Среднее значение:  38.11
Мода:  38.0
Медиана:  38.18717948717948
Максимальное значение:  40.8
Минимальное значение:  35.4
Размах:  5.399999999999999
Среднее квадратическое отклонение:  0.6552257639408873
Дисперсия:  0.4293208017319193
Q1 - первый квартиль:  37.9
Q3 - третий квартиль:  38.5
Межквартильный размах:  0.6000000000000014


**пульс**

In [73]:
# сгруппируем данные по исходу и операциям.
print(data.groupby(['surgery', 'outcome'])['pulse'].mean())
print(data.groupby(['surgery'])['pulse'].mean())
print(data.groupby(['outcome'])['pulse'].mean())

surgery  outcome
1.0      1.0        68.715909
         2.0        88.183673
         3.0        80.037037
2.0      1.0        57.448718
         2.0        84.470588
         3.0        84.625000
Name: pulse, dtype: float64
surgery
1.0    76.396341
2.0    65.504505
Name: pulse, dtype: float64
outcome
1.0    63.421687
2.0    87.227273
3.0    81.744186
Name: pulse, dtype: float64


У группы лошадок, попавших под нож хирурга, пульс, в среднем, был выше, чем у тех, до кого доктор не добрался. 
Лошадки с более высоким пульсом умирали (или их усыпляли), а с меньшим - выживали.
В общем, наверное, эти данные можно использовать для заполнения пропусков. У нас совсем немного изменились  среднее арифметическое и медиана и уменьшились меры разброса. 

In [74]:
data_without_nan['pulse'].fillna(data.groupby(['surgery', 'outcome'])['pulse'].transform('mean'), inplace=True)

In [75]:
print('Пульс до преобразований')
statistics(data['pulse'])
print('-------')
print('Пульс после заполнения пропусков')
statistics(data_without_nan['pulse'])

Пульс до преобразований
Среднее значение:  71.91304347826087
Мода:  48.0
Медиана:  64.0
Максимальное значение:  184.0
Минимальное значение:  30.0
Размах:  154.0
Среднее квадратическое отклонение:  28.630556660735003
Дисперсия:  819.7087747035575
Q1 - первый квартиль:  48.0
Q3 - третий квартиль:  88.0
Межквартильный размах:  40.0
-------
Пульс после заполнения пропусков
Среднее значение:  72.20333333333333
Мода:  48.0
Медиана:  66.0
Максимальное значение:  184.0
Минимальное значение:  30.0
Размах:  154.0
Среднее квадратическое отклонение:  27.697022361367473
Дисперсия:  767.1250476860899
Q1 - первый квартиль:  48.0
Q3 - третий квартиль:  88.18367346938776
Межквартильный размах:  40.183673469387756


**Частота дыхания**

In [76]:
# группируем по категориям. 
print(data.groupby(['outcome', 'surgery'])['resp_rate'].mean())
print(data.groupby(['surgery'])['resp_rate'].mean())
print(data.groupby(['outcome'])['resp_rate'].mean())

outcome  surgery
1.0      1.0        32.500000
         2.0        25.830986
2.0      1.0        34.111111
         2.0        29.307692
3.0      1.0        34.826087
         2.0        24.615385
Name: resp_rate, dtype: float64
surgery
1.0    33.375000
2.0    26.134021
Name: resp_rate, dtype: float64
outcome
1.0    29.278912
2.0    33.034483
3.0    31.138889
Name: resp_rate, dtype: float64


Хирурги предпочитали резать лошадок, которые часто-часто дышали. У выживших лошадок кол-во вдохов/мин было чуть меньше, чем у тех, кто не пережил госпитализацию.  
После замены данных среднее арифметическое осталось без изменений, медиана стала больше, а данные - более однородными (размах уменьшился).

In [77]:
data_without_nan['resp_rate'].fillna(data.groupby(['surgery', 'outcome'])['resp_rate'].transform('mean'), inplace=True)

In [78]:
print('Кол-во вдохов/мин до преобразований')
statistics(data['resp_rate'])
print('-------')
print('Кол-во вдохов/мин после заполнения пропусков')
statistics(data_without_nan['resp_rate'])

Кол-во вдохов/мин до преобразований
Среднее значение:  30.417355371900825
Мода:  20.0
Медиана:  24.5
Максимальное значение:  96.0
Минимальное значение:  8.0
Размах:  88.0
Среднее квадратическое отклонение:  17.642231385134664
Дисперсия:  311.24832824663054
Q1 - первый квартиль:  18.5
Q3 - третий квартиль:  36.0
Межквартильный размах:  17.5
-------
Кол-во вдохов/мин после заполнения пропусков
Среднее значение:  30.456666666666667
Мода:  32.0
Медиана:  28.0
Максимальное значение:  96.0
Минимальное значение:  8.0
Размах:  88.0
Среднее квадратическое отклонение:  15.915901027969948
Дисперсия:  253.31590553213485
Q1 - первый квартиль:  20.0
Q3 - третий квартиль:  34.869565217391305
Межквартильный размах:  14.869565217391305


In [81]:
data_without_nan.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
surgery      300 non-null float64
age          300 non-null int64
rect_temp    300 non-null float64
pulse        300 non-null float64
resp_rate    300 non-null float64
extr_temp    300 non-null object
pain         300 non-null object
outcome      300 non-null float64
dtypes: float64(5), int64(1), object(2)
memory usage: 18.9+ KB


**Датафрейм без пропусков**

In [80]:
data_without_nan

surgery  age  rect_temp  pulse  resp_rate extr_temp     pain  outcome
0        2.0    1  38.500000   66.0       28.0         3        5      2.0
1        1.0    1  39.200000   88.0       20.0   no_info        3      3.0
2        2.0    1  38.300000   40.0       24.0         1        3      1.0
3        1.0    9  39.100000  164.0       84.0         4        2      2.0
4        2.0    1  37.300000  104.0       35.0   no_info  no_info      2.0
..       ...  ...        ...    ...        ...       ...      ...      ...
295      1.0    1  38.052381  120.0       70.0         4        2      3.0
296      2.0    1  37.200000   72.0       24.0         3        4      3.0
297      1.0    1  37.500000   72.0       30.0         4        4      2.0
298      1.0    1  36.500000  100.0       24.0         3        3      1.0
299      1.0    1  37.200000   40.0       20.0   no_info  no_info      3.0

[300 rows x 8 columns]